### PR TITLE
Fix handling of recursive bounds in constraints

### PR DESCRIPTION
**Pull request by smarter**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala [hunk 1]
```diff
@@ -87,6 +87,10 @@ trait ConstraintHandling[AbstractContext] {
 
   protected def addOneBound(param: TypeParamRef, bound: Type, isUpper: Boolean)(using AbstractContext): Boolean =
     if !constraint.contains(param) then true
+    else if !isUpper && param.occursIn(bound)
+      // We don't allow recursive lower bounds when defining a type,
+      // so we shouldn't allow them as constraints either.
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala:92]: I am not 100% sure about performance implications here. Let's test this.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
Yes, that looks like a much better way to fix #8976

**Comment by smarter** [comment]
I think it's within the usual variance but we can run again to check

**Comment by smarter** [comment]
test performance please

**Pull request merged**